source/particle/generator/probability_density_function.cc

```diff
@@ -0,0 +1,295 @@
+      {
+        // Get the local accumulated probabilities for every cell
+        const std::vector<double> accumulated_cell_weights = local_accumulated_cell_weights();
+
+        // Sum the local integrals over all nodes
+        double local_function_integral = accumulated_cell_weights.back();
+        const double global_function_integral = Utilities::MPI::sum (local_function_integral,
+                                                                     this->get_mpi_communicator());
+
+        AssertThrow(global_function_integral > std::numeric_limits<double>::min(),
+                    ExcMessage("The integral of the user prescribed probability density function over the domain equals zero, "
+                               "ASPECT has no way to determine the cell of generated tracers. Please ensure that the provided "
+                               "function is positive in at least a part of the domain, also check the syntax of the function."));
+
+        // Determine the starting weight of this process, which is the sum of
+        // the weights of all processes with a lower rank
+        double start_weight = 0.0;
+        MPI_Scan(&local_function_integral, &start_weight, 1, MPI_DOUBLE, MPI_SUM, this->get_mpi_communicator());
+        start_weight -= local_function_integral;
+
+        // Build the map between integrated probability density and cell
+        std::map<double, types::LevelInd> cell_weights;
+        typename DoFHandler<dim>::active_cell_iterator cell = this->get_dof_handler().begin_active(),
+                                                       endc = this->get_dof_handler().end();
+        for (unsigned int i = 0; cell!=endc; ++cell)
+          if (cell->is_locally_owned())
+            {
+              // adjust the cell_weights according to the local starting weight
+              cell_weights.insert(std::make_pair(accumulated_cell_weights[i] + start_weight, types::LevelInd(cell->level(),cell->index())));
+              ++i;
+            }
+
+        // Calculate start and end IDs so there are no gaps
+        const types::particle_index start_id = llround(n_tracers * start_weight / global_function_integral) + 1;
+
+        return generate_particles_in_subdomain(cell_weights,global_function_integral,start_weight,n_tracers, 1.1 * start_id);
+      }
+
+      template <int dim>
+      std::vector<double>
+      ProbabilityDensityFunction<dim>::local_accumulated_cell_weights () const
+      {
+        //evaluate function at all cell midpoints, sort cells according to weight
+        const QMidpoint<dim> quadrature_formula;
+        const unsigned int n_quadrature_points = quadrature_formula.size();
+
+        FEValues<dim> fe_values (this->get_mapping(),
+                                 this->get_fe(),
+                                 quadrature_formula,
+                                 update_quadrature_points |
+                                 update_JxW_values);
```

Comment:
At first, I thought that we don't actually need a `FEValues` object here, because `position[q]` is always equal to `cell->center()` and `JxW(q)` is always `cell->measure()`. But that's not true because you use the mapping. Can you add a comment here to this effect, explaining why you need a `FEValues` object to begin with?